### ISSUE TITLE
How can i get full image paths?

### ISSUE THREAD

**Issue by jesperlandberg** [agentejo/cockpit#816]
Is there any good way of getting the full path to images (rest api)? I'm not hosting cockpit on the same server as the front-end.

**Comment by nisdec** [comment]
I have the same question.

**Comment by michabbb** [comment]
@jesperlandberg ah, you mean path on the server, not url to open that image in the browser, right ? 😏

**Comment by jesperlandberg** [comment]
@michabbb correct=)

**Comment by michabbb** [comment]
@jesperlandberg pardon, misunderstanding 🙈 😏

**Comment by nisdec** [comment]
Since this cms is mainly used as a "headless" cms, it would just make sense to serve absolute image paths, because more often than not is this cms hosted on a different domain than the frontend. 
Even if this would be an opt-in option.

@michabbb Your suggestion might be okay in a "smaller" frontend environment, but it becomes a pain when you have to deal with a ton of components, services and maybe even multiple apps, which all get their data from this cms.
To feed a "image_url"-variable through everything is just a unnecessary pain.

**Comment by michabbb** [comment]
@nisdec i am just evaluating this cms at the moment, thats why i am interested in some issues here. as i said, sorry - misunderstanding here 😏 beside the fact that there exists NULL docu about "howto write an addon", i really like what i see, so far. and yes, i understand you.

**Comment by gryphonmyers** [comment]
It's an oft-requested feature https://github.com/agentejo/cockpit/issues/695 and one we've argued with @aheinze about before https://github.com/agentejo/cockpit/issues/470

**Comment by raffaelj** [comment]
Create a file `/config/bootstrap.php` with this content:

```php
<?php

$app->on('collections.save.before.collectionname', function($name, &$entry, $isUpdate) use ($app) {

    $asset_fields = [];
    
    // get collection schema
    $collection = $app->module('collections')->collection($name);
    
    // find asset fields
    foreach ($collection['fields'] as $field) {
        if ($field['type'] == 'asset')
            $asset_fields[] = $field['name'];
    }
    
    foreach ($asset_fields as $key => $val) {
        
        $entry[$val]['absolute_path'] = $app['site_url'] . $app['base_url'] . '/storage/uploads' . $entry[$val]['path'];
        
        $entry[$val]['real_path'] = COCKPIT_DIR . '/storage/uploads' . $entry[$val]['path'];
    }
    
});
```
or `$app->on('collections.save.before', function(...` for all collections.

Now all assets have `absolute_path` and `real_path`, when they are saved.

The trigger `collections.find.before` doesn't allow to modify data, so it's necessary to create the path when saving the entry.

**Comment by nisdec** [comment]
@raffaelj Thank you for your solution!

Unfortunately I get the following error when I try your solution:
`Warning: Parameter 1 to LimeExtra\App::{closure}() expected to be a reference, value given in /cms/lib/Lime/App.php on line 592`

Any idea?

**Comment by raffaelj** [comment]
@nisdec I can't reproduce it. Can you explain, what you did before? Did you use the assets manager or image-field or asset-field or the api...? Maybe you have a typo...

Do you use the (latest) next branch or master?
Is your PHP version > 7.0?

**Comment by nisdec** [comment]
@raffaelj Okay I managed to solve the error simply by updating Cockpit, sorry. I was on PHP 7.2 with Cockpit 0.6.0 or 0.6.1. Now I'm on PHP 7.2 with Cockpit 0.8.0.

It actually works now when I use an asset-field! Is there also a way to do the same for an image-field?

Thank you so much.

**Comment by raffaelj** [comment]
@nisdec 
> Is there also a way to do the same for an image-field?

I'm not sure. After skimming the code, I don't think so. Image uploads are handled different than assets. It looks like they are just simple files without any database entries.


You could create a custom api endpoint:
```php
<?php
/*
  copy this code to `/config/api/imagepath.php`
  
  call it with `/api/imagepath?token=xxtokenxx`
  json body: {"src":"imageid","h":"100"}
  
*/

$thumb_url = $this->invoke('\\Cockpit\\Controller\\RestApi', 'image');

$relative_path = str_replace($this['site_url'].$this['base_route'], '', $thumb_url);
        
$real_path = COCKPIT_DIR . $relative_path;

return [
    'thumb_url' => $thumb_url,
    'relative_path' => $relative_path,
    'real_path' => $real_path
];
```

**Comment by annda** [comment]
I have a site that consumes Cockpit content via API and I would like to insert images in HTML fields, using uploaded assets. The WYSIWYG editor truncates the image paths and outputs relative links, like '/storage/...'. 

In a headless CMS I would expect full paths in HTML, like  'http(s)/:.mysite.domain/storage....'

Is there a way to configure the editor to prepend the site path to src in image tags?

I found this option in the TinyMCE docs https://www.tiny.cloud/docs/configure/file-image-upload/#images_upload_base_path but I don't know how, or whether, it is possible to configure Cockpit to use it.

The alternative would be to parse all HTML client-side and replace the image links, which I would rather avoid.

**Comment by annda** [comment]
I absolutely agree that storing full paths in the database is nonsense and would cause more harm than good.

But I am not sure what is wrong with the API itself prepending the URL on database output. Cockpit already uses `site_url` in configuration, and the paths to assets, uploads, etc. are also separately defined and can be overridden on a per-system basis. 

Basically, my wish for the future is that the Cockpit **API** would process the storage paths internally and output the URLs in links. That way the files could even be stored on another system. And the clients would only call the API and remain agnostic of which assets are stored where. But that is probably a whole new feature request.

In the meantime, I will stick to a simple regex replacement in the client.

**Comment by jasonday** [comment]
> But I am not sure what is wrong with the API itself prepending the URL on database output. Cockpit already uses `site_url` in configuration, and the paths to assets, uploads, etc. are also separately defined and can be overridden on a per-system basis.

I agree. Additionally, Cockpit is even more aggressive in that if you put in a fully qualified url, cockpit will rewrite it to a relative url - which is a huge pain - https://github.com/agentejo/cockpit/issues/1102

**Comment by darkylmnx** [comment]
There should be an option for absolute URL's as this is really annoying on JAMstack or SPA apps. On these kind of apps, the CMS is always (98% of the time) on a different domain so the absolute path would be the most accurate solution.
